### DISCUSSION TITLE
Serial read not reading the entire package in one transaction

### DISCUSSION THREAD

**Discussion by rawinza555** [nasa/fprime#1065] [Q&A] [answered]
Hi. I have a STM32 with a RF module attached connecting to a raspberry pi via USB port. I used LinuxSerialDriver to communicate between the two devices. The way I set it up is to have a stm32 to handle RF transaction and put whatever it reads to R Pi via USB port. However, it seems like that the data sent was chopped.

To verify that it's not a RF issue, I have a STM32 sending a array of 1-10 via serial port and, as expected, the packet received on R Pi is partially chopped at a random spot at each transaction. I'm wondering if there's a parameter in LinuxSerialDriver that I could adjust this to read the entire package in one go? any help would be appreciated.

Btw, I used baud 115200 and no parity bit on both devices.

**Comment by timcanham** [comment, marked as answer]
It is actually common that the receiving end of a UART transaction will chop packets at different points. This is usually due to buffering in the hardware. You should just structure your logic to assume this will happen and reassemble the packet at a higher level.

**Reply by rawinza555** [reply]
Thanks Tim. I did that and it works great.

As a follow up question, what would be the best approach to passive component using LinuxSerialDriver? I know that I need a preamble function to send buffer beforehand but I couldn't make preamble function to run on a passive component (possibly due to it not having a thread?).

**Comment by timcanham** [comment]
Yes, the `preamble()` function is only called if the component is active, and it's meant to be used for one-time startup code when the thread is first started. I wouldn't use it for sending UART data. What were you planning on doing in the preamble if you had it? Why are you using a passive component to receive UART data?

**Reply by rawinza555** [reply]
Preamble will be used to send buffer to LinuxSerialDriver prior to receive the data via UART.

I created a component in F' that act as a interface between a small MCU (STM32 in this case) that handle downlink/uplink RF communication and a raspberry pi. UART will be used for 

1.) send downlink data  to MCU which will be then sent via a RF module
2.) read data from MCU, which is from what RF module received.

There are two identical STM32 (Nucleo H7) with SPIRIT1 transceiver. One connected to a R Pi and the other connected to a PC running a OpenMCT. The purpose is to test OpenMCT on receiving TM and sending TC. This system will eventually replace F' GDS but I think I might want both of them to work concurrently at this stage so I have something to compare to in real time. 

As for why passive component, I'm just trying to explore the possibility. If it wouldn't work then there's no problem having it as an active component.
